mysys/crc32/crc32_arm64.c

```diff
@@ -121,7 +129,7 @@ asm(".arch_extension crypto");
 	PREF4X64L2(buffer,(PREF_OFFSET), 8) \
 	PREF4X64L2(buffer,(PREF_OFFSET), 12)
 
-
+int pmull_aarch64_runtime_flag = 0;
```

Comment:
This should be declared `static`, as requested elsewhere. Please remove the redundant to zero-initialization. Global or `static` variables will be zero-initialized by default in the BSS.

I might even call this variable `pmull_supported` for improved readability. Given that it should be specific to this compilation unit, there is no need to include "aarch64" in the name.

Reply:
Agreed to rename this variable to `pmull_supported` for improved readability and remove zero-initialization here, thanks.

But `pmull_supported` is initialized in `ut_crc32_init` by calling `crc32c_aarch64_available`.
For `Arm64 PMULL` and `crc32 Arm64 instructions` were fused into `crc32c_aarch64` , `pmull_supported` is used as a switch to enable/disable `Arm64 PMULL` in `crc32c_aarch64`.  IMHO, it's better to declare `pmull_supported` as `extern `not `static `here.

Reply:
I also suggested to merge the two functions `crc32_aarch64_available()` and `crc32c_aarch_available()` to one function that would initialize the `pmull_supported` flag. If you think that it is cleaner to avoid changing `crc32_aarch64_available()` at all, and to perform this check in a separate `crc32c_aarch64_available()` function, then I think that that function should return the `const char*` message that I suggested. I do not see any need to export `pmull_supported` outside the compilation unit.

Reply:
`pmull_supported` is the runtime check flag in `crc32c_aarch64`(mysys/crc32/crc32_arm64.c) which determine whether skip `pmull `or not.

`crc32_aarch64_available` would be called in `ut_crc32_init`(storage/innobase/ut/ut0crc32.cc) where `pmull_supported `is initialized.
It's required `pmull_supported ` should be extern linkage.

Reply:
I meant that in `ut_crc32_init()` we can do the following:
```diff
diff --git a/storage/innobase/ut/ut0crc32.cc b/storage/innobase/ut/ut0crc32.cc
--- a/storage/innobase/ut/ut0crc32.cc
+++ b/storage/innobase/ut/ut0crc32.cc
@@ -342,11 +342,11 @@ allocations, would not hurt if called twice, but would be pointless. */
 void ut_crc32_init()
 {
 #ifndef HAVE_CRC32_VPMSUM
-# if defined(__GNUC__) && defined(HAVE_ARMV8_CRC) && defined(HAVE_ARMV8_CRYPTO)
-  if (crc32c_aarch64_available())
+# if defined(__GNUC__) && defined(HAVE_ARMV8_CRC)
+  if (const char *crc32c_implementation= crc32c_aarch64_available())
   {
     ut_crc32_low= crc32c_aarch64;
-    ut_crc32_implementation= "Using ARMv8 crc32 + pmull instructions";
+    ut_crc32_implementation= crc32c_implementation;
     return;
   }
 # elif defined(TRY_SSE4_2)
```
In this way, the `pmull_supported` can be declared `static` in `crc32_arm64.c`.

Reply:
Yep, got it, thanks.